python/ray/data/read_api.py

```diff
@@ -2900,6 +2901,54 @@ def from_torch(
     )
 
 
+@PublicAPI
+def from_lancedb(
+    *,
+    uri: str,
+    columns: Optional[List[str]] = None,
+    filter: Optional[str] = None,
+    parallelism: int = -1,
+    ray_remote_args: Optional[Dict[str, Any]] = None,
+) -> Dataset:
+    """
+    Create a :class:`~ray.data.Dataset` from a LanceDB dataset. The dataset to read from
+    is specified using a fully qualified ```uri```. Using LanceDB, any
+    intended columns or filters are applied, and the files that satisfy
+    the query are distributed across Ray read tasks. The number of tasks is
+    determined by ``parallelism`` which can be requested from this interface or
+    automatically chosen if unspecified (see the``parallelism`` arg below).
+
+    Examples:
+        >>> import ray
+        >>> ds = ray.data.read_lancedb(
+        ...     uri="./db_name.lance",
+        ...     columns=["column_name", "label"],
+        ...     filter="label = 2 AND column_name IS NOT NULL",
+        ...     parallelism=64
+        ... )
+
+    Args:
+        uri: The URI of the source LanceDB dataset to read from.
+            Currently supports local file paths, S3, or GCS URIs are supported.
+        columns: The columns to read from the dataset.
+            If not specified, all columns are read.
+        filter: The filter to apply to the dataset.
+            If not specified, no filter is applied.
+        parallelism: Degree of parallelism to use for the Dataset
```

Comment:
Can Ray work with datasets that do their own parallelism?  For example, you could read a fragment with `batch_readahead` and lance will do multi-threading on its own.  This is a bit more efficient than kicking off a bunch of "read_batch" tasks since we only have to read/parse/decode the metadata a single time instead of multiple times.

Reply:
Yes, within each Ray task (per-fragment), multi-threading is allowed.